File: python/06_002.py
def merge_sort(arr):
   """Sort array using merge sort algorithm."""
   # Base case
   if len(arr) <= 1:
      return arr
   
   # Divide input array
   mid = len(arr) // 2
   left = arr[:mid]
   right = arr[mid:]
   
   # Recursively sort halves
   left = merge_sort(left)
   right = merge_sort(right)
   
   # Merge sorted halves
   return merge(left, right)

def merge(left, right):
   """Merge two sorted arrays."""
   result = []
   i = j = 0
   
   # Compare and merge elements
   while i < len(left) and j < len(right):
      if left[i] <= right[j]:
         result.append(left[i])
         i += 1
      else:
         result.append(right[j])
         j += 1
   
   # Append remaining elements
   result.extend(left[i:])
   result.extend(right[j:])
   return result

# Demonstrate merge sort
if __name__ == "__main__":
   # Test cases
   test_cases = [
      [23, 16, 6, 59, 3, 11, 37],
      [1, 2, 3, 4, 5],
      [5, 4, 3, 2, 1],
      []
   ]
   
   for case in test_cases:
      orig = case.copy()
      sorted_case = merge_sort(case)
      print(f"Original: {orig}")
      print(f"Sorted:   {sorted_case}\n")

File: cpp/06_002.cpp
#include <iostream>
#include <vector>

void merge(std::vector<int>& array, int left, int mid, 
           int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;
    std::vector<int> leftArr(n1), rightArr(n2);
    
    // Copy data to temporary arrays
    for (int i = 0; i < n1; i++) 
        leftArr[i] = array[left + i];
    for (int j = 0; j < n2; j++) 
        rightArr[j] = array[mid + 1 + j];
    
    int i = 0, j = 0, k = left;
    
    // Merge the temporary arrays back into the original array
    while (i < n1 && j < n2) {
        if (leftArr[i] <= rightArr[j]) {
            array[k++] = leftArr[i++];
        } else {
            array[k++] = rightArr[j++];
        }
    }
    
    // Copy remaining elements of leftArr
    while (i < n1) 
        array[k++] = leftArr[i++];
    
    // Copy remaining elements of rightArr
    while (j < n2) 
        array[k++] = rightArr[j++];
}

void mergeSort(std::vector<int>& array, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(array, left, mid);
        mergeSort(array, mid + 1, right);
        merge(array, left, mid, right);
    }
}

int main() {
    std::vector<int> array = {23, 16, 6, 59, 3, 11, 37};
    mergeSort(array, 0, array.size() - 1);
    
    for (int num : array) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
    
    return 0;
}